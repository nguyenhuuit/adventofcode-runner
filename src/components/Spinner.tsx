import React, {useState, useEffect} from 'react';
import {Text} from 'ink';
import spinners from 'cli-spinners';
import type {SpinnerName} from 'cli-spinners';

type Props = {
	type?: SpinnerName;
  loading?: Boolean;
};

function Spinner({type = 'dots', loading = false}: Props) {
	const [frame, setFrame] = useState(0);
	const spinner = spinners[type];

	useEffect(() => {
		if (loading) {
			const timer = setInterval(() => {
				setFrame(previousFrame => {
					const isLastFrame = previousFrame === spinner.frames.length - 1;
					return isLastFrame ? 0 : previousFrame + 1;
				});
			}, spinner.interval);
	
			return () => {
				clearInterval(timer);
			};
		}
		return () => {}
	}, [spinner, loading]);

	return loading ? <Text>{spinner.frames[frame]}</Text> : <Text>{spinner.frames[0]}</Text>;
}

export default Spinner;