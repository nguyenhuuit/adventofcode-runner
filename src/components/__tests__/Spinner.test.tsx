import { render } from '@testing-library/react';

import React, { act } from 'react';

import Spinner from '../Spinner';

interface MockTextProps {
  children: React.ReactNode;
}

jest.mock('ink', () => ({
  Text: ({ children }: MockTextProps) => <span data-testid="text">{children}</span>,
}));

jest.mock('cli-spinners', () => ({
  dots: {
    frames: ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'],
    interval: 80,
  },
  line: {
    frames: ['-', '\\', '|', '/'],
    interval: 130,
  },
}));

describe('Spinner', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should render first frame when not loading', () => {
    const { container } = render(<Spinner loading={false} />);
    expect(container).toMatchSnapshot();
  });

  it('should render first frame when loading is undefined', () => {
    const { container } = render(<Spinner />);
    expect(container).toMatchSnapshot();
  });

  it('should animate when loading is true', () => {
    const { container } = render(<Spinner loading={true} />);

    expect(container).toMatchSnapshot();

    act(() => {
      jest.advanceTimersByTime(80);
    });

    expect(container).toMatchSnapshot();

    act(() => {
      jest.advanceTimersByTime(80);
    });

    expect(container).toMatchSnapshot();
  });

  it('should loop back to first frame after last frame', () => {
    const { container } = render(<Spinner loading={true} />);

    for (let i = 0; i < 10; i++) {
      act(() => {
        jest.advanceTimersByTime(80);
      });
    }

    expect(container).toMatchSnapshot();
  });

  it('should use different spinner type', () => {
    const { container } = render(<Spinner type="line" loading={true} />);

    expect(container).toMatchSnapshot();

    act(() => {
      jest.advanceTimersByTime(130);
    });

    expect(container).toMatchSnapshot();
  });

  it('should stop animation when loading becomes false', () => {
    const { container, rerender } = render(<Spinner loading={true} />);

    act(() => {
      jest.advanceTimersByTime(80);
    });
    expect(container).toMatchSnapshot();

    rerender(<Spinner loading={false} />);

    expect(container).toMatchSnapshot();

    act(() => {
      jest.advanceTimersByTime(80);
    });
    expect(container).toMatchSnapshot();
  });

  it('should restart animation when loading becomes true again', () => {
    const { container, rerender } = render(<Spinner loading={false} />);

    expect(container).toMatchSnapshot();

    rerender(<Spinner loading={true} />);

    act(() => {
      jest.advanceTimersByTime(80);
    });

    expect(container).toMatchSnapshot();
  });

  it('should handle multiple timer advances', () => {
    const { container } = render(<Spinner loading={true} />);

    act(() => {
      jest.advanceTimersByTime(240); // 3 intervals
    });

    expect(container).toMatchSnapshot();
  });

  it('should use default type when type is not specified', () => {
    const { container } = render(<Spinner loading={true} />);
    expect(container).toMatchSnapshot();
  });

  it('should handle rapid loading state changes', () => {
    const { container, rerender } = render(<Spinner loading={true} />);

    act(() => {
      jest.advanceTimersByTime(80);
    });
    expect(container).toMatchSnapshot();

    rerender(<Spinner loading={false} />);
    rerender(<Spinner loading={true} />);

    expect(container).toMatchSnapshot();
  });
});
