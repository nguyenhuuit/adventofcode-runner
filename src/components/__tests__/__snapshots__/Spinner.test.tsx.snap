// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner should animate when loading is true 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠋
  </span>
</div>
`;

exports[`Spinner should animate when loading is true 2`] = `
<div>
  <span
    data-testid="text"
  >
    ⠙
  </span>
</div>
`;

exports[`Spinner should animate when loading is true 3`] = `
<div>
  <span
    data-testid="text"
  >
    ⠹
  </span>
</div>
`;

exports[`Spinner should handle multiple timer advances 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠸
  </span>
</div>
`;

exports[`Spinner should handle rapid loading state changes 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠙
  </span>
</div>
`;

exports[`Spinner should handle rapid loading state changes 2`] = `
<div>
  <span
    data-testid="text"
  >
    ⠙
  </span>
</div>
`;

exports[`Spinner should loop back to first frame after last frame 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠋
  </span>
</div>
`;

exports[`Spinner should render first frame when loading is undefined 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠋
  </span>
</div>
`;

exports[`Spinner should render first frame when not loading 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠋
  </span>
</div>
`;

exports[`Spinner should restart animation when loading becomes true again 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠋
  </span>
</div>
`;

exports[`Spinner should restart animation when loading becomes true again 2`] = `
<div>
  <span
    data-testid="text"
  >
    ⠙
  </span>
</div>
`;

exports[`Spinner should stop animation when loading becomes false 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠙
  </span>
</div>
`;

exports[`Spinner should stop animation when loading becomes false 2`] = `
<div>
  <span
    data-testid="text"
  >
    ⠋
  </span>
</div>
`;

exports[`Spinner should stop animation when loading becomes false 3`] = `
<div>
  <span
    data-testid="text"
  >
    ⠋
  </span>
</div>
`;

exports[`Spinner should use default type when type is not specified 1`] = `
<div>
  <span
    data-testid="text"
  >
    ⠋
  </span>
</div>
`;

exports[`Spinner should use different spinner type 1`] = `
<div>
  <span
    data-testid="text"
  >
    -
  </span>
</div>
`;

exports[`Spinner should use different spinner type 2`] = `
<div>
  <span
    data-testid="text"
  >
    \\
  </span>
</div>
`;
